src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java

```diff
@@ -324,79 +326,93 @@ public void setBaseUrl(String url) {
-                            //Don't send pings if we're using long polling.
-                            if (transportEnum != TransportEnum.LONG_POLLING) {
-                                activatePingTimer();
+            negotiate.flatMapCompletable(negotiateResponse -> {
+                logger.debug("Starting HubConnection.");
+                if (transport == null) {
+                    Single<String> tokenProvider = negotiateResponse.getAccessToken() != null ? Single.just(negotiateResponse.getAccessToken()) : accessTokenProvider;
+                    switch (transportEnum) {
+                        case LONG_POLLING:
+                            transport = new LongPollingTransport(localHeaders, httpClient, tokenProvider);
+                            break;
+                        default:
+                            transport = new WebSocketTransport(localHeaders, httpClient);
+                    }
+                }
+
+                transport.setOnReceive(this.callback);
+                transport.setOnClose((message) -> stopConnection(message));
+
+                return transport.start(negotiateResponse.getFinalUrl()).andThen(Completable.defer(() -> {
+                    String handshake = HandshakeProtocol.createHandshakeRequestMessage(
+                            new HandshakeRequestMessage(protocol.getName(), protocol.getVersion()));
+
+                    connectionState = new ConnectionState(this);
+
+                    return transport.send(handshake).andThen(Completable.defer(() -> {
+                        timeoutHandshakeResponse(handshakeResponseTimeout, TimeUnit.MILLISECONDS);
+                        return handshakeResponseSubject.andThen(Completable.defer(() -> {
+                            hubConnectionStateLock.lock();
+                            try {
+                                hubConnectionState = HubConnectionState.CONNECTED;
+                                logger.info("HubConnection started.");
+                                resetServerTimeout();
+                                //Don't send pings if we're using long polling.
+                                if (transportEnum != TransportEnum.LONG_POLLING) {
+                                    activatePingTimer();
+                                }
+                            } finally {
+                                hubConnectionStateLock.unlock();
                             }
-                        } finally {
-                            hubConnectionStateLock.unlock();
-                        }
 
-                        return Completable.complete();
+                            return Completable.complete();
+                        }));
                     }));
                 }));
-            }));
-        // subscribe makes this a "hot" completable so this runs immediately
-        }).subscribeWith(start);
+            // subscribe makes this a "hot" completable so this runs immediately
+            }).subscribe(() -> {
```

Comment:
The CONNECTING->CONNECTED state transition is hidden in this diff, but we should assert the previous state where we can.

Reply:
I have a PR that extracts this into a method, I'll add the logging though